File: server/controllers/conversationController.ts
import { Request, Response } from 'express';
import { postgresService } from '../services/postgresService.ts';
import { cacheService } from '../services/cacheService.ts';
import { aiDutchTutorService } from '../services/aiDutchTutorService.ts';
import { SCENARIOS } from '../services/scenarioData.ts';
import { ConversationTurn } from '../types/index.ts';
import { logger } from '../services/logger.ts';
import { metricsService } from '../services/metricsService.ts';

// Active SSE client connections map
const sseClients: Map<string, Response[]> = new Map();

export const startSession = async (req: Request, res: Response) => {
  const { scenarioId, userId = 'usr_java_engineer_1' } = req.body;
  const scenario = SCENARIOS.find((s) => s.id === scenarioId) || SCENARIOS[0];

  const session = await postgresService.createSession(userId, scenario.id);

  // Initialize initial tutor turn in session
  const initialTurn: ConversationTurn = {
    id: `turn_${Date.now()}_init`,
    sessionId: session.sessionId,
    sender: 'tutor',
    dutchText: scenario.initialMessageNl,
    englishTranslation: 'Welcome! Let us begin our conversation.',
    timestamp: new Date().toISOString(),
    b2Upgrade: 'Tip: Luister aandachtig naar de vraag en gebruik inversie als je antwoordt met "Vandaag..." of "Nu...".',
    pronunciationTip: 'Spreek rustig en let op de beklemtoning van samengestelde woorden.',
  };

  await postgresService.addConversationTurn(initialTurn);

  // Cache in Redis for high concurrency
  const cacheKey = `nl:session:${userId}:${session.sessionId}:turns`;
  await cacheService.set(cacheKey, [initialTurn], 3600);
  metricsService.incrementActiveSessions();

  return res.json({
    session,
    scenario,
    initialTurn,
  });
};

export const getSessionHistory = async (req: Request, res: Response) => {
  const { sessionId } = req.params;
  const session = await postgresService.getSession(sessionId);
  if (!session) {
    return res.status(404).json({ error: 'Session not found' });
  }

  // Try Redis cache first
  const cacheKey = `nl:session:${session.userId}:${sessionId}:turns`;
  const cachedTurns = await cacheService.get<ConversationTurn[]>(cacheKey);
  if (cachedTurns) {
    return res.json({ turns: cachedTurns, cached: true });
  }

  const turns = await postgresService.getSessionHistory(sessionId);
  await cacheService.set(cacheKey, turns, 3600);
  return res.json({ turns, cached: false });
};

export const handleSseStream = (req: Request, res: Response) => {
  const { sessionId } = req.params;

  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache, no-transform');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();

  // Send connected ping
  res.write(`data: ${JSON.stringify({ type: 'CONNECTED', sessionId })}\n\n`);

  const existing = sseClients.get(sessionId) || [];
  existing.push(res);
  sseClients.set(sessionId, existing);

  req.on('close', () => {
    const clients = sseClients.get(sessionId) || [];
    sseClients.set(
      sessionId,
      clients.filter((c) => c !== res)
    );
  });
};

function cleanSpokenDuplicates(text: string): string {
  if (!text) return '';
  const words = text.trim().replace(/\s+/g, ' ').split(' ');
  const deduped: string[] = [];
  for (let i = 0; i < words.length; i++) {
    const current = words[i];
    const prev = deduped[deduped.length - 1];
    const currentClean = current.toLowerCase().replace(/[.,!?;:]/g, '');
    const prevClean = prev ? prev.toLowerCase().replace(/[.,!?;:]/g, '') : '';
    if (prev && currentClean && currentClean === prevClean) {
      continue;
    }
    deduped.push(current);
  }
  return deduped.join(' ');
}

export const sendMessage = async (req: Request, res: Response) => {
  try {
    const { sessionId } = req.params;
    const { text, userId = 'usr_java_engineer_1' } = req.body;

    if (!text || !text.trim()) {
      return res.status(400).json({ error: 'Message text is required' });
    }

    const session = await postgresService.getSession(sessionId);
    if (!session) {
      return res.status(404).json({ error: 'Session not found' });
    }

    const scenario = SCENARIOS.find((s) => s.id === session.scenarioId) || SCENARIOS[0];
    const history = await postgresService.getSessionHistory(sessionId);

    // Sanitize any consecutive speech recognition duplications (e.g. "Hallo Hallo" -> "Hallo")
    const cleanedText = cleanSpokenDuplicates(text);

    // 1. Create student turn
    const studentTurn: ConversationTurn = {
      id: `turn_${Date.now()}_usr`,
      sessionId,
      sender: 'user',
      dutchText: cleanedText,
      timestamp: new Date().toISOString(),
    };
    await postgresService.addConversationTurn(studentTurn);

    // Broadcast to SSE stream that student message was accepted
    broadcastSse(sessionId, { type: 'STUDENT_MESSAGE', turn: studentTurn });

    // 2. Evaluate with Gemini AI Dutch Tutor
    const evaluation = await aiDutchTutorService.evaluateAndReply(cleanedText, scenario, history);

    // 3. Create AI Tutor reply turn with live corrections and grammar evaluation
    const tutorTurn: ConversationTurn = {
      id: `turn_${Date.now()}_ttr`,
      sessionId,
      sender: 'tutor',
      dutchText: evaluation.replyInDutch,
      englishTranslation: evaluation.englishTranslation,
      timestamp: new Date().toISOString(),
      grammarStatus: evaluation.grammarStatus,
      grammarSummary: evaluation.grammarSummary,
      improvedDutch: evaluation.improvedDutch,
      corrections: evaluation.corrections,
      b2Upgrade: evaluation.b2Upgrade,
      pronunciationTip: evaluation.pronunciationTip,
      xpEarned: evaluation.xpEarned,
    };
    await postgresService.addConversationTurn(tutorTurn);

    // 4. Update user XP and streak in PostgreSQL
    const updatedUser = await postgresService.addXpAndStreak(session.userId, evaluation.xpEarned);

    // 5. Update Redis cache
    const cacheKey = `nl:session:${session.userId}:${sessionId}:turns`;
    const allTurns = await postgresService.getSessionHistory(sessionId);
    await cacheService.set(cacheKey, allTurns, 3600);

    // Broadcast tutor reply to SSE stream
    broadcastSse(sessionId, {
      type: 'TUTOR_REPLY',
      turn: tutorTurn,
      suggestedQuickReplies: evaluation.suggestedQuickReplies,
      user: updatedUser,
    });

    return res.json({
      studentTurn,
      tutorTurn,
      suggestedQuickReplies: evaluation.suggestedQuickReplies,
      user: updatedUser,
    });
  } catch (error) {
    logger.error('Error handling message in conversationController:', { error: String(error) });
    return res.status(500).json({ error: 'Interne fout bij het verwerken van het bericht.' });
  }
};

function broadcastSse(sessionId: string, data: unknown) {
  const clients = sseClients.get(sessionId) || [];
  for (const client of clients) {
    try {
      client.write(`data: ${JSON.stringify(data)}\n\n`);
    } catch (err) {
      logger.error('Failed to write to SSE client', { error: String(err) });
    }
  }
}

File: backend/src/main/java/com/praatnederlands/controller/ConversationController.java
package com.praatnederlands.controller;

import com.praatnederlands.model.ConversationSession;
import com.praatnederlands.model.ConversationTurn;
import com.praatnederlands.model.DutchTutorResponse;
import com.praatnederlands.model.RoleplayScenario;
import com.praatnederlands.service.AiDutchTutorService;
import com.praatnederlands.service.ScenarioService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/api/conversation")
public class ConversationController {

    private final ScenarioService scenarioService;
    private final AiDutchTutorService tutorService;

    private final Map<String, ConversationSession> sessions = new ConcurrentHashMap<>();
    private final Map<String, List<ConversationTurn>> sessionHistory = new ConcurrentHashMap<>();

    public record StartSessionRequest(String scenarioId, String userId) {}
    public record SendMessageRequest(String text, String userId) {}

    public ConversationController(ScenarioService scenarioService, AiDutchTutorService tutorService) {
        this.scenarioService = scenarioService;
        this.tutorService = tutorService;
    }

    @PostMapping("/session")
    public ResponseEntity<Map<String, Object>> startSession(@RequestBody(required = false) StartSessionRequest req) {
        String scenarioId = (req != null && req.scenarioId() != null) ? req.scenarioId() : "sollicitatie";
        String userId = (req != null && req.userId() != null) ? req.userId() : "usr_java_engineer_1";

        RoleplayScenario scenario = scenarioService.getScenario(scenarioId);
        String sessionId = "sess_" + System.currentTimeMillis() + "_" + UUID.randomUUID().toString().substring(0, 7);

        ConversationSession session = new ConversationSession(
            sessionId,
            userId,
            scenarioId,
            Instant.now().toString(),
            Instant.now().toString(),
            1
        );
        sessions.put(sessionId, session);

        ConversationTurn initialTurn = new ConversationTurn(
            "turn_" + System.currentTimeMillis() + "_init",
            sessionId,
            "tutor",
            scenario.initialMessageNl(),
            "Welcome! Let us begin our conversation.",
            Instant.now().toString(),
            null, null, null, null,
            "Tip: Luister aandachtig naar de vraag en gebruik inversie als je antwoordt.",
            "Spreek rustig en let op de beklemtoning van samengestelde woorden.",
            null
        );

        List<ConversationTurn> turns = new ArrayList<>();
        turns.add(initialTurn);
        sessionHistory.put(sessionId, turns);

        return ResponseEntity.ok(Map.of(
            "session", session,
            "scenario", scenario,
            "initialTurn", initialTurn
        ));
    }

    @GetMapping("/session/{sessionId}/history")
    public ResponseEntity<Map<String, Object>> getHistory(@PathVariable String sessionId) {
        List<ConversationTurn> turns = sessionHistory.getOrDefault(sessionId, List.of());
        return ResponseEntity.ok(Map.of("turns", turns, "cached", false));
    }

    @PostMapping("/session/{sessionId}/message")
    public ResponseEntity<?> sendMessage(
        @PathVariable String sessionId,
        @RequestBody SendMessageRequest req
    ) {
        if (req == null || req.text() == null || req.text().isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("error", "Message text is required"));
        }

        ConversationSession session = sessions.get(sessionId);
        if (session == null) {
            return ResponseEntity.status(404).body(Map.of("error", "Session not found"));
        }

        RoleplayScenario scenario = scenarioService.getScenario(session.scenarioId());
        List<ConversationTurn> history = sessionHistory.computeIfAbsent(sessionId, k -> new ArrayList<>());

        // 1. Record student turn
        ConversationTurn studentTurn = new ConversationTurn(
            "turn_" + System.currentTimeMillis() + "_usr",
            sessionId,
            "user",
            req.text().trim(),
            null,
            Instant.now().toString(),
            null, null, null, null, null, null, null
        );
        history.add(studentTurn);

        // 2. Evaluate with Dutch tutor engine (Priority RAG enabled)
        DutchTutorResponse evaluation = tutorService.evaluateAndReply(req.text().trim(), scenario, history);

        // 3. Record tutor response turn
        ConversationTurn tutorTurn = new ConversationTurn(
            "turn_" + System.currentTimeMillis() + "_ttr",
            sessionId,
            "tutor",
            evaluation.replyInDutch(),
            evaluation.englishTranslation(),
            Instant.now().toString(),
            evaluation.grammarStatus(),
            evaluation.grammarSummary(),
            evaluation.improvedDutch(),
            evaluation.corrections(),
            evaluation.b2Upgrade(),
            evaluation.pronunciationTip(),
            evaluation.xpEarned()
        );
        history.add(tutorTurn);

        return ResponseEntity.ok(Map.of(
            "studentTurn", studentTurn,
            "tutorTurn", tutorTurn,
            "suggestedQuickReplies", evaluation.suggestedQuickReplies() != null ? evaluation.suggestedQuickReplies() : List.of()
        ));
    }
}
